### PR TITLE
Enable support for MessagePack

### PR DESCRIPTION
**Pull request by Y-Sindo**
## Description

Enable support for MessagePack hub protocol.

Usage:
* Option 1: Enable MessagePack in the `Startup` class:
    ```cs
        public override void Configure(IFunctionsHostBuilder builder)
	    {
	        builder.Services.Configure<SignalROptions>(o =>
	        {
	            // o.ServiceTransportType = ServiceTransportType.Persistent;
	            o.MessagePackHubProtocol = new MessagePackHubProtocol();
	        });
	    }
    ```
* Option 2: Add a setting in the application settings
    ```
    Azure:SignalR:HubProtocol:MessagePack:Enabled=true
    ```
---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] Please add REST spec PR link to the SDK PR
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/SignalROptions.cs [hunk 1]
```diff
@@ -35,6 +39,21 @@ public class SignalROptions : IOptionsFormatter
...
         public ObjectSerializer JsonObjectSerializer { get; set; }
 
+        /// <summary>
+        /// Gets or sets the MessagePack hub <see cref="IHubProtocol"/>. Defaults to null and MessagePack hub protocol is not used.
+        /// </summary>
+        public IHubProtocol MessagePackHubProtocol
+        {
+            get => _messagePackHubProtocol; set
+            {
+                if (value != null && value.Name.ToLowerInvariant() != "messagepack")
```
Comment (vicancy) [inline sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/SignalROptions.cs:49]: use string.Equal

File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Constants.cs [hunk 2]
```diff
@@ -9,6 +9,8 @@ internal static class Constants
         public const string AzureSignalREndpoints = "Azure:SignalR:Endpoints";
         public const string AzureSignalRHubProtocol = "Azure:SignalR:HubProtocol";
         public static string AzureSignalRNewtonsoftCamelCase = $"{AzureSignalRHubProtocol}:NewtonsoftJson:CamelCase";
+        public static string AzureSignalRMessagePackHubProtocol = $"{AzureSignalRHubProtocol}:MessagePack";
+        public const string Enabled = "enabled";
```
Comment (JialinXin) [inline sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Constants.cs:13]: Shall we make it boolean?

File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Config/ServiceManagerStore.cs [hunk 3]
```diff
@@ -57,6 +58,17 @@ private IInternalServiceHubContextStore CreateHubContextStore(string connectionS
                 // Serves as a reload token provider only
                 .WithConfiguration(new EmptyConfiguration(_configuration))
                 .WithCallingAssembly();
+
+            if (_options.MessagePackHubProtocol != null)
+            {
+                serviceManagerBuilder.AddHubProtocol(_options.MessagePackHubProtocol);
+            }
+            // Allow isolated-process runtimes such as JS, C#-isolated to enable MessagePack hub protocol
+            else if (string.Equals(_configuration[Constants.AzureSignalRMessagePackHubProtocol], Constants.Enabled, StringComparison.InvariantCultureIgnoreCase))
```
Comment (vicancy) [inline sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Config/ServiceManagerStore.cs:67]: OrdinalIgnoreCase is enough?

Follow similar configuration pattern as what kestrel do?
```
"Https": {
    "Enabled": false,
}
```

File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Constants.cs (hunk 2 above)
Comment (Y-Sindo) [reply sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Constants.cs:13]: updated

File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Config/ServiceManagerStore.cs (hunk 3 above)
Comment (Y-Sindo) [reply sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Config/ServiceManagerStore.cs]: updated

File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/SignalROptions.cs (hunk 1 above)
Comment (Y-Sindo) [reply sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/SignalROptions.cs:49]: fixed

**Review by JialinXin — APPROVED** [review]

File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Config/ServiceManagerStore.cs [hunk 4]
```diff
@@ -57,6 +58,17 @@ private IInternalServiceHubContextStore CreateHubContextStore(string connectionS
                 // Serves as a reload token provider only
                 .WithConfiguration(new EmptyConfiguration(_configuration))
                 .WithCallingAssembly();
+
+            if (_options.MessagePackHubProtocol != null)
+            {
+                serviceManagerBuilder.AddHubProtocol(_options.MessagePackHubProtocol);
+            }
+            // Allow isolated-process runtimes such as JS, C#-isolated to enable MessagePack hub protocol
+            else if (string.Equals(_configuration[Constants.AzureSignalRMessagePackHubProtocolEnabled], bool.TrueString, StringComparison.OrdinalIgnoreCase))
```
Comment (vicancy) [inline sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Config/ServiceManagerStore.cs:67]: use `bool.TryParse`

File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/SignalROptions.cs (hunk 1 above)
Comment (vicancy) [reply sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/SignalROptions.cs:49]: still seeing the old version

File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Config/ServiceManagerStore.cs (hunk 4 above)
Comment (Y-Sindo) [reply sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Config/ServiceManagerStore.cs]: updated

File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/SignalROptions.cs (hunk 1 above)
Comment (Y-Sindo) [reply sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/SignalROptions.cs]: Updated

**Review by JoshLove-msft — APPROVED** [review]

**Review by vicancy — APPROVED** [review]

**Review by KrzysztofCwalina — APPROVED** [review]

**Comment by Y-Sindo** [comment]
#auto-merge

**Pull request merged**